### PR TITLE
[feature] support more queries with range (>=, <= etc)

### PR DESCRIPTION
**Pull request by Dingying0410**
### Summary 
This PR is to support the queries that have range in it. such as '>=', '<=', etc., and also combine the `NotEqual` query we had in #62 since they can all be merged into the `NodeHasComparator` condition in this PR.

Regarding the rspec, I created several shared examples so we can test multiple similar behaviors. 

### Details 
Previously we decided to not to cache any queries with these comparison operators, however, queries with both comparison operators and other bound queries can be cached. E.g
```
Site.where(a: 1).where('b > ?' 1) can be cached based on a as dependency 
Site.where('b > ?' 1) should not be cached
```

Also, as according to [here](https://piechowski.io/post/how-to-use-greater-than-less-than-active-record/), there are two forms of queries with comparator: 
1. Site.where(a: 1).where('b > ?' 1) -- this will be analyzed as `Arel::Nodes::Grouping` when we parse the children of the wheres, thus, we look at `node.expr` of it. `node.expr` is usually a `Arel::Nodes::SqlLiteral` in this case
2. Site.where(a: 1).where(b: 2..Float::INFINITY) -- this will be analyzed as `Arel::Nodes::LessThan` when we parse the children of the wheres, thus, we categorize `Arel::Nodes::LessThan` and also other node class as `NodeHasComparators`

My current approach is to return `bind_params` when we parse these statements 
- for queries with the only a comparator, we return an empty object 
- for queries with the comparator and other bound conditions, it will involve those conditions and cache

### Test plan
- rspec for >=, >, <=, <

### TODO: 
remove the duplicate Arel::Nodes::NotEqual block in #62 since they are brought in this PR

### CODE REVIEW

File: lib/redis_memo/memoize_query/cached_select.rb
Comment (katyho) [inline lib/redis_memo/memoize_query/cached_select.rb:382]: Wouldn't NotEqual already be covered [here](https://github.com/chanzuckerberg/redis-memo/blob/main/lib/redis_memo/memoize_query/cached_select.rb#L340)? If so, then we can probably remove the condition with just `Arel::Nodes::NotEqual` if we have this condition

**Comment by donaldong** [comment]
The build failure seems legit btw

File: lib/redis_memo/memoize_query/cached_select.rb
Comment (Dingying0410) [reply lib/redis_memo/memoize_query/cached_select.rb:382]: oh good catch thanks! I forgot to mention that line 341 can be removed after this will be merged.

Comment (katyho) [reply lib/redis_memo/memoize_query/cached_select.rb:383]: Why do we need to wait for this to be merged in order to remove the code on line 341? Couldn't we remove it as part of this PR?

Comment (Dingying0410) [reply lib/redis_memo/memoize_query/cached_select.rb:377]: Yup I missed this option, thanks : )

File: lib/redis_memo/memoize_query/cached_select.rb
Comment (Dingying0410) [inline lib/redis_memo/memoize_query/cached_select.rb:340]: removing since this will be supported in line 377 in this PR

File: spec/memoize_query_spec.rb
Comment (Dingying0410) [inline spec/memoize_query_spec.rb:446]: moving to the bottom of this file since they are all range queries relations

File: spec/memoize_query_spec.rb
Comment (Dingying0410) [inline spec/memoize_query_spec.rb:546]: using `if defined?` since not every context needs all there relations to test

File: spec/memoize_query_spec.rb
Comment (katyho) [reply spec/memoize_query_spec.rb:650]: Also, these tests mostly look the same, the only difference being the operator `<, <=, >, >=`, so it looks like there's a lot of code duplication. I feel like it'd be sufficient to just have a list of 1. relations with only comparators, and  2. relations with comparator and bound conditions, and combine most of these test cases together.

File: spec/memoize_query_spec.rb
Comment (Dingying0410) [inline spec/memoize_query_spec.rb:608]: This will still be considered as only having comparators since it will be parsed as `Arel::Nodes::Grouping`, so will not be cached.

File: spec/memoize_query_spec.rb
Comment (Dingying0410) [reply spec/memoize_query_spec.rb]: > This test doesn't contain only unbound conditions right?

and that actually means "contain unbound", I updated the title though

> 1. relations with only comparators, and 2. relations with comparator and bound conditions

updated in each context

>  combine most of these test cases together

I combined the cases for > with >= and < with <=, didn't combine all of them into one since there is nuance in the update and create action (used to test the results are valid after update) for different operators. I was able to resolve the code dup. For whether we should test all cases, i still think that we should test each test case as possible, to everyone sure it works correctly. I think it does not matter too much to just test one more relation.

File: spec/memoize_query_spec.rb
Comment (Dingying0410) [reply spec/memoize_query_spec.rb]: Good point. I refactored the shared examples a bit, but did not end up doing `relations_with_only_one_comparator.each do |relation|` (passing the relation) because of some implementation details for `let`. but I passed a list of `relations` as a let variable.

File: spec/memoize_query_spec.rb
Comment (Dingying0410) [reply spec/memoize_query_spec.rb]: Hmm I think this is the implementation details for each test/ context so it seems weird to put that into the abstract example.

File: spec/memoize_query_spec.rb
Comment (Dingying0410) [reply spec/memoize_query_spec.rb]: replied in https://github.com/chanzuckerberg/redis-memo/pull/63#discussion_r608168350

File: lib/redis_memo/memoize_query/cached_select.rb
Comment (donaldong) [inline lib/redis_memo/memoize_query/cached_select.rb:377]: Can we move this to L318?

File: lib/redis_memo/memoize_query/cached_select.rb
Comment (donaldong) [inline lib/redis_memo/memoize_query/cached_select.rb:350]:
```suggestion
   %w[< >= < <= !=].include?(expr.to_s)
```
at some point, we should add some rubocop lint rules to this repo :)

Comment (Dingying0410) [reply lib/redis_memo/memoize_query/cached_select.rb:350]: I think this is a reverse relationship, but I used the %w

File: lib/redis_memo/memoize_query/cached_select.rb
Comment (donaldong) [inline lib/redis_memo/memoize_query/cached_select.rb:349]: Do you mean
```suggestion
    %w[> >= < <= !=].include?(expr.to_s)
```

File: spec/memoize_query_spec.rb
Comment (donaldong) [inline spec/memoize_query_spec.rb:748]: We should probably use `reload` all the time right? the test cases might not run in order.

File: spec/memoize_query_spec.rb
Comment (donaldong) [inline spec/memoize_query_spec.rb:637]: You probably want this to be a function. Currently it's just a variable named `update`

File: spec/memoize_query_spec.rb
Comment (donaldong) [inline spec/memoize_query_spec.rb:639]: Why do we want to separate this out from the main test case?

Also, looks like it's defined again later with the same content.

**Review by donaldong — APPROVED** [review]
Thanks! Posted some comments

In general, I think the new test cases are not super readable -- but I guess it does have good coverage,  so approving.

File: lib/redis_memo/memoize_query/cached_select.rb
Comment (donaldong) [inline lib/redis_memo/memoize_query/cached_select.rb:316]: before,  when an expr is a Arel::Nodes::SqlLiteral, we'd reject the query.

Currently, we accept the query it the string literal contains `comparison` operator. (note there are edge cases such as "a = '<='")

when the expr does not have a comparison operator, we should still reject the query instead of continuing.

**Review by donaldong — CHANGES_REQUESTED** [review]
Actually, I found an issue after revisiting it.

File: spec/memoize_query_spec.rb
Comment (Dingying0410) [reply spec/memoize_query_spec.rb:639]: this does not appear in the new commit

File: spec/memoize_query_spec.rb
Comment (Dingying0410) [reply spec/memoize_query_spec.rb:637]: this has been removed in new commit

File: lib/redis_memo/memoize_query/cached_select.rb
Comment (Dingying0410) [reply lib/redis_memo/memoize_query/cached_select.rb:316]: ah good catch, thanks

File: lib/redis_memo/memoize_query/cached_select.rb
Comment (Dingying0410) [reply lib/redis_memo/memoize_query/cached_select.rb:349]: this has been removed, but it was the case that `expr` is `a > 3` so we wanted to see if the `expr` contains comparators in the list

**Review by donaldong — APPROVED** [review]
🎉

**Pull request merged**
